### PR TITLE
fix for php7.1

### PR DESCRIPTION
**Pull request by rafmichalak84**
this is fix to issue #1389 
and incompatible method returnString for zendEngine3

I tested the rebuild phalcon 3.1.x it works well for now

php.7.1.1 and php.7.2.0-dev
zephir 0.9.6a-dev-72dbb2063e

### CODE REVIEW

**Comment by Jurigag** [comment]
Well will check tomorrow in work, build it with this fix and check all tests.

**Comment by rafmichalak84** [comment]
There is one test with PHP warning
ArithmeticTest.php
line:264 $this->assertSame($this->class->letStatementVarMinus(""), -"");

left side and right side is OK, return int(0)
but PHP warning create error in test
when error_reporting set to 0, all PASS

**Comment by Jurigag** [comment]
Well it could be good to test also all phalcon tests with this fix.

**Comment by sergeyklay** [comment]
Could you please cover your changes by test?

**Comment by rafmichalak84** [comment]
Yes all zephir unit tests is 100% OK
Tomorrow I check phalcon tests

**Comment by rafmichalak84** [comment]
Status
there is one in my opinion important bug when static method in class one call another static method in class two then throw Exception call to undefined method

Now I look in gdb for scope calls

rest is fine

------
However, multiple static calls inside the class

```
class class2 {
        public static function _call3() -> string {
                return "1234-5678";
        }

        public static function _call2() -> string {
                return self::_call3();
        }

        public static function call1() -> string {
                return self::_call2();
        }
}
```
```
<?php
var_dump(\Comtest\Class2::call1());
```
```
PHP Fatal error:  Uncaught RuntimeException: Call to undefined method self::_call3(), did you mean 'undefined'? in /home/raf/comtest/test.php:3
Stack trace:
#0 [internal function]: Comtest\class2::_call2()
#1 /home/raf/comtest/test.php(3): Comtest\class2::call1()
#2 {main}
  thrown in /home/raf/comtest/test.php on line 3
```

File: Library/Backends/ZendEngine3/Backend.php [hunk 1]
```diff
@@ -287,9 +287,9 @@ public function declareConstant($type, $name, $value, CompilationContext $contex
         }
 
         if ($dType == 'null') {
-            $context->codePrinter->output('zephir_declare_class_constant_null(' . $ce . ', SL("' . $name . '"));');
+            $context->codePrinter->output('zend_declare_class_constant_null(' . $ce . ', SL("' . $name . '"));');
```
Comment (Jurigag) [inline Library/Backends/ZendEngine3/Backend.php]: Why exactly change this to zend stuff? Isn't better to fix broken zephir code?

File: Library/Backends/ZendEngine3/Backend.php [hunk 2]
```diff
@@ -287,9 +287,9 @@ public function declareConstant($type, $name, $value, CompilationContext $contex
...
         } else {
-            $context->codePrinter->output('zephir_declare_class_constant_' . $dType . '(' . $ce . ', SL("' . $name . '"), ' . $value . ');');
+            $context->codePrinter->output('zend_declare_class_constant_' . $dType . '(' . $ce . ', SL("' . $name . '"), ' . $value . ');');
```
Comment (Jurigag) [inline Library/Backends/ZendEngine3/Backend.php]: Why exactly change this to zend stuff? Isn't better to fix broken zephir code?

File: kernels/ZendEngine3/extended/fcall.c [hunk 3]
```diff
@@ -1077,14 +1082,14 @@ int zephir_call_function_opt(zend_fcall_info *fci, zend_fcall_info_cache *fci_ca
 		}
 
 		if (!info) {
-			char *callable_name;
-			if (!zephir_is_callable_ex(&fci->function_name, fci->object, IS_CALLABLE_CHECK_SILENT, &callable_name, NULL, fci_cache, &error)) {
+			zend_string *callable_name;
+			if (!zend_is_callable_ex(&fci->function_name, fci->object, IS_CALLABLE_CHECK_SILENT, &callable_name, fci_cache, &error)) {
```
Comment (Jurigag) [inline kernels/ZendEngine3/extended/fcall.c]: Again. Why exactly change to zend stuff? Isn't better to fix zephir related stuff?

File: Library/Backends/ZendEngine3/Backend.php (hunk 1 above)
Comment (rafmichalak84) [reply Library/Backends/ZendEngine3/Backend.php]: OK you right

File: kernels/ZendEngine3/extended/fcall.c (hunk 3 above)
Comment (rafmichalak84) [reply kernels/ZendEngine3/extended/fcall.c]: Yes you right

**Comment by rafmichalak84** [comment]
The current result
In some cases (i think when set scope) is don't access in object to private method (sometimes not always)
When I find it in gdb i check scope setting.

Error 1,2,4 is the same

There were 4 errors:

1) DispatcherMvcEventsTest::testEvents
Error: Cannot access protected property PHPUnit_Framework_Constraint_IsEqual::$exporter

/home/rafal/NetBeansProjects/cphalcon/unit-tests/DispatcherMvcEventsTest.php:74
/home/rafal/NetBeansProjects/cphalcon/unit-tests/DispatcherMvcEventsTest.php:205

2) DispatcherMvcEventsTest::testBeforeExceptionEvent
Error: Cannot access protected property PHPUnit_Framework_Constraint_IsEqual::$exporter

/home/rafal/NetBeansProjects/cphalcon/unit-tests/DispatcherMvcEventsTest.php:74
/home/rafal/NetBeansProjects/cphalcon/unit-tests/DispatcherMvcEventsTest.php:295

3) ModelsCalculationsTest::testCalculationsMysql
Error: Cannot instantiate abstract class Phalcon\Mvc\Model

/home/rafal/NetBeansProjects/cphalcon/unit-tests/ModelsCalculationsTest.php:131
/home/rafal/NetBeansProjects/cphalcon/unit-tests/ModelsCalculationsTest.php:77

4) ViewEnginesVoltTest::testVoltEngineBuiltInFunctions
Error: Cannot access private property SomeObject::$_data

/home/rafal/NetBeansProjects/cphalcon/unit-tests/ViewEnginesVoltTest.php:44
/home/rafal/NetBeansProjects/cphalcon/unit-tests/views/test11/index.volt.php:2
/home/rafal/NetBeansProjects/cphalcon/unit-tests/ViewEnginesVoltTest.php:195

ERRORS!
Tests: 132, Assertions: 7726, Errors: 4, Skipped: 11.

**Comment by Jurigag** [comment]
For me it seems like php problem cuz:

`Error: Cannot access protected property PHPUnit_Framework_Constraint_IsEqual::$exporter`

Is not zephir or phalcon thing.

`Error: Cannot instantiate abstract class Phalcon\Mvc\Model`

I already created issue about this.

File: Library/Backends/ZendEngine3/Backend.php [hunk 4]
```diff
@@ -362,7 +362,7 @@ public function assignZval(Variable $variable, $code, CompilationContext $contex
         }
     }
 
-    public function returnString($value, CompilationContext $context, $useCodePrinter = true)
+    public function returnString($value, CompilationContext $context, $useCodePrinter = true, $doCopy = true)
```
Comment (sergeyklay) [inline Library/Backends/ZendEngine3/Backend.php]: Could you please explain this?

Comment (rafmichalak84) [inline Library/Backends/ZendEngine3/Backend.php]: Yes because ZendEngine3 extends ZendEngine2 where returnString is with additional parameters

**Comment by sergeyklay** [comment]
@steffengy Could you please take a look

**Comment by dreamsxin** [comment]
Can clean multi commit merge to one.

**Comment by rafmichalak84** [comment]
Yes of course I merge to one
Currently I looking for line where zephir include php from outside and lossing (scope to object) property access visibility.

Cannot access to private or propetected property. This is very imortant because programmers usually use this access modifires. I want to spend all weekend and i think find it ;-)

**Comment by Jurigag** [comment]
Yea gl with that. Im not even sure this is zephir problem tbh cuz 

`Error: Cannot access protected property PHPUnit_Framework_Constraint_IsEqual::$exporter` doesn't happens on first occurence or first test. It happens on xxx occurence. Like few tests before this it's used without any problems.

Though it happens each time in same test on same line which is weird.

**Comment by dreamsxin** [comment]
How to reproduce this error?
I see the Travis CI build succes.

**Comment by Jurigag** [comment]
It's happening for phalcon tests. Not zephir tests. You have to clone his repo, checkout fix-php7.1 compile phalcon using his zephir witch changes and run phalcon tests.

**Comment by dreamsxin** [comment]
Reference method zend_call_function, 
When change the following code  test in php7.1.1 it's ok:
```diff
--- a/kernels/ZendEngine3/extended/fcall.c
+++ b/kernels/ZendEngine3/extended/fcall.c
@@ -1177,28 +1177,16 @@ int zephir_call_function_opt(zend_fcall_info *fci, zend_fcall_info_cache *fci_ca
                ZVAL_COPY_VALUE(param, arg);
        }
 
-#if PHP_VERSION_ID >= 70100
-       EG(fake_scope) = calling_scope;
-#else
-       EG(scope) = calling_scope;
-#endif
+       if (UNEXPECTED(func->op_array.fn_flags & ZEND_ACC_CLOSURE)) {
+               ZEND_ASSERT(GC_TYPE((zend_object*)func->op_array.prototype) == IS_OBJECT);
+               GC_REFCOUNT((zend_object*)func->op_array.prototype)++;
+               ZEND_ADD_CALL_FLAG(call, ZEND_CALL_CLOSURE);
+       }
 
        if (func->type == ZEND_USER_FUNCTION) {
                int call_via_handler = (func->common.fn_flags & ZEND_ACC_CALL_VIA_TRAMPOLINE) != 0;
 
 #if PHP_VERSION_ID >= 70100
-               EG(fake_scope) = func->common.scope;
-               call->symbol_table = NULL;
-#else
-               EG(scope) = func->common.scope;
-               call->symbol_table = fci->symbol_table;
-#endif
-               if (UNEXPECTED(func->op_array.fn_flags & ZEND_ACC_CLOSURE)) {
-                       ZEND_ASSERT(GC_TYPE((zend_object*)func->op_array.prototype) == IS_OBJECT);
-                       GC_REFCOUNT((zend_object*)func->op_array.prototype)++;
-                       ZEND_ADD_CALL_FLAG(call, ZEND_CALL_CLOSURE);
-               }
-#if PHP_VERSION_ID >= 70100
                zend_init_execute_data(call, &func->op_array, fci->retval);
                zend_execute_ex(call);
 #else
@@ -1216,13 +1204,6 @@ int zephir_call_function_opt(zend_fcall_info *fci, zend_fcall_info_cache *fci_ca
        } else if (func->type == ZEND_INTERNAL_FUNCTION) {
                int call_via_handler = (func->common.fn_flags & ZEND_ACC_CALL_VIA_TRAMPOLINE) != 0;
                ZVAL_NULL(fci->retval);
-               if (func->common.scope) {
-#if PHP_VERSION_ID >= 70100
-                       EG(fake_scope) = func->common.scope;
-#else
-                       EG(scope) = func->common.scope;
-#endif
-               }
                call->prev_execute_data = EG(current_execute_data);
                call->return_value = NULL; /* this is not a constructor call */
                EG(current_execute_data) = call;
@@ -1276,11 +1257,6 @@ int zephir_call_function_opt(zend_fcall_info *fci, zend_fcall_info_cache *fci_ca
                }
        }
 
-#if PHP_VERSION_ID >= 70100
-       EG(fake_scope) = orig_scope;
-#else
-       EG(scope) = orig_scope;
-#endif
        zend_vm_stack_free_call_frame(call);
 
        if (EG(current_execute_data) == &dummy_execute_data) {
diff --git a/kernels/ZendEngine3/fcall.c b/kernels/ZendEngine3/fcall.c
index 1357752..9cc44d8 100644
--- a/kernels/ZendEngine3/fcall.c
+++ b/kernels/ZendEngine3/fcall.c
@@ -417,14 +417,6 @@ int zephir_call_user_function(zval *object_pp, zend_class_entry *obj_ce, zephir_
                }
        }
 
-       if (obj_ce) {
-#if PHP_VERSION_ID >= 70100
-               EG(fake_scope) = obj_ce;
-#else
-               EG(scope) = obj_ce;
-#endif
-       }
-
        if (!cache_entry || !*cache_entry) {
                if (zephir_globals_ptr->cache_enabled) {
 
@@ -493,12 +485,6 @@ int zephir_call_user_function(zval *object_pp, zend_class_entry *obj_ce, zephir_
        //status = ZEPHIR_ZEND_CALL_FUNCTION_WRAPPER(&fci, &fcic, info);
        status = zephir_call_function_opt(&fci, &fcic, info, params);
 
-#if PHP_VERSION_ID >= 70100
-               EG(fake_scope) = old_scope;
-#else
-               EG(scope) = old_scope;
-#endif
-
```

**Comment by dreamsxin** [comment]
Use Cache, the second will happen can,t get called class
```shell
PHP_METHOD(Test_Oo, testInstance9) {
...
        ZEPHIR_CALL_CE_STATIC(&o, test_oo_oodynamica_ce, "getnew", &_0, 49);
...
}

PHP_METHOD(Test_Oo, testInstance10) {
...
        ZEPHIR_CALL_CE_STATIC(&o, test_oo_oodynamicb_ce, "getnew", &_0, 49);
...
}
```

Change:
```diff
diff --git a/kernels/ZendEngine3/fcall.c b/kernels/ZendEngine3/fcall.c
index 7a7e944..3986050 100644
--- a/kernels/ZendEngine3/fcall.c
+++ b/kernels/ZendEngine3/fcall.c
@@ -474,6 +474,7 @@ int zephir_call_user_function(zval *object_pp, zend_class_entry *obj_ce, zephir_
 #else
                fcic.function_handler = *cache_entry;
 #endif
+               fcic.called_scope = obj_ce ? obj_ce : Z_OBJCE(EG(current_execute_data)->This);
        }
```

But php-7.0 test failed.

**Comment by rafmichalak84** [comment]
OK I see
in some place is losing call(ing/ed) scope and reference only / (empty class and namespace)

------
OK thanks @dreamsxin but I have the same error, please check on your version maybe we have different code modification

```
namespace Comtest;

class class3 {
        public static function call2() {
                var className, fullClassName;
                let className = get_called_class();
                let fullClassName = "\\" . className;
                return new {fullClassName}();
        }

        public static function call1() {
                var o;
                let o = \Comtest\Class3::call2();
                return o;
        }
}
```
and test
```
var_dump(\Comtest\Class3::call2());
var_dump(\Comtest\Class3::call1());
```
I have
```
object(Comtest\class3)#1 (0) {
}
PHP Fatal error:  Class '\' not found in /home/rafal/NetBeansProjects/x/comtest/test.php on line 10

Fatal error: Class '\' not found in /home/rafal/NetBeansProjects/x/comtest/test.php on line 10
```

In the place when I add your code
```
fcic.called_scope =  obj_ce ?  obj_ce : Z_OBJCE(EG(current_execute_data)->This);
```
I checked by gdb and my obj_ce and Z_OBJCE return empty object

**Comment by dreamsxin** [comment]
My test branch https://github.com/dreamsxin/phalcon7-zephir/blob/master/kernels/ZendEngine3/

**Comment by rafmichalak84** [comment]
@dreamsxin thanks a lot
when I use your version I have
in zephir unit-test 100% pass
in phalcon unit-test segmentation fault

something goes wrong
but after comment line
```
fcic.called_scope = obj_ce ? obj_ce : Z_OBJCE(EG(current_execute_data)->This);
```
I have a lot Phalcon\Mvc\Model\Exception: Model 'Phalcon\Mvc\Model' could not be loaded

**Comment by dreamsxin** [comment]
I test in PHP 7.1.1, And not full test,  May be check       EG(current_execute_data):
```c
EG(current_execute_data) ? Z_OBJ(EG(current_execute_data)->This) : NULL
```

**Comment by rafmichalak84** [comment]
@dreamsxin 
I find one solution to "Class '\' not found"
```diff
extended/fcall.c ~line1200
+       if(fci_cache->object) {
+             Z_OBJ(call->This) = fci_cache->object;
+       }
	if (UNEXPECTED(func->op_array.fn_flags & ZEND_ACC_CLOSURE)) {
		ZEND_ASSERT(GC_TYPE((zend_object*)func->op_array.prototype) == IS_OBJECT);
		GC_REFCOUNT((zend_object*)func->op_array.prototype)++;
		ZEND_ADD_CALL_FLAG(call, ZEND_CALL_CLOSURE);
	}

	if (func->type == ZEND_USER_FUNCTION) {
```

At this moment I have all pass zephir unit-tests and most phalcon
```
F.FF.SS....SSSSS....S......................................EEE.  63 / 129 ( 48%)
.....WF.F.....EEF.EEEEF.F...F.FF...F.F...F.......E.....SSS..... 126 / 129 ( 97%)
...                                                             129 / 129 (100%)

top errors

1) ModelsCriteriaTest::testModelsMysql
Phalcon\Mvc\Model\Exception: At least one model is required to build the query

6) ModelsRelationsTest::testModelsMysql
BadMethodCallException: Wrong number of parameters

/home/rafal/NetBeansProjects/cphalcon/unit-tests/ModelsRelationsTest.php:146
/home/rafal/NetBeansProjects/cphalcon/unit-tests/ModelsRelationsTest.php:75

```

**Comment by dreamsxin** [comment]
I suggest trying use `zend_call_function` replace `zephir_call_user_function` reduce complexity.

**Comment by Jurigag** [comment]
But it's for some reason - i guess it's debugging co you can compile zephir extension with debugging symbols and have more information than we would just have all zend calls.

**Comment by dreamsxin** [comment]
But there is a big difference bewteen PHP-7.0 and PHP-7.1. Can only be re combed to find out the problem.

**Comment by Jurigag** [comment]
Still - it should be fixed on zephir side. Just use `#if #elif #else #endif` to have code base for php 7.1, 7.0 and 5.6, it's obvious i think.

**Comment by Jurigag** [comment]
@rafmichalak84 just as soon as you will fix any kind of issue like this visibility thing or maybe https://github.com/phalcon/zephir/issues/1392 then mention me so i can compile zephir and check all phalcon tests again.

**Comment by rafmichalak84** [comment]
@Jurigag Zephir all tests are green, there is one leak: example in test of the foreign keys in phalcon

and in function zephir_get_called_class
called_scope->name has "Cannot access memory at address" like a not assigned reference

call_stack
Phalcon_Mvc_Model__groupResult
Phalcon_Mvc_Model_count
Phalcon_Mvc_Model__checkForeignKeysRestrict
Phalcon_Mvc_Model__preSave
Phalcon_Mvc_Model_save

example in file
[a.php.txt](https://github.com/phalcon/zephir/files/777096/a.php.txt)

**Comment by Jurigag** [comment]
Yea i know about this problem, i created issue, not sure if it's this pr or just zephir master even has this problem? Didn't check this.

Still zephir tests are zephir tests, phalcon tests are covering MUCH MORE zephir stuff. Like i mean if this is fixed - https://github.com/phalcon/zephir/pull/1390#issuecomment-277809280

Also what's going with php 7.0 tests? Not green.

**Comment by rafmichalak84** [comment]
@Jurigag 
I enabled old code for 7.0 and new for 7.1 by #if and all zephir tests now are OK

**Comment by Jurigag** [comment]
Okay, let me pull changes and i will check phalcon tests.

**Comment by Jurigag** [comment]
Oh right there is `zephir_get_called_class` seg fault, you wrote it above, then i wait for fix for this, let me know when you will handle this.

**Comment by rafmichalak84** [comment]
Give me your coredump ;-) it's very helpfull to fix it

**Comment by Jurigag** [comment]
http://pastebin.com/XqvpwNGL

Didn't have this a few commits earlier on this PR, there were just https://github.com/phalcon/zephir/issues/1392 this problem.

**Comment by rafmichalak84** [comment]
Now I must for 3-5 days forget this fix
Today I found some interesting moments which I think are the key
when I change fci_cache->object to fci->object in extended/fcall.c line
```
call = zend_vm_stack_push_call_frame(ZEND_CALL_TOP_FUNCTION, func, fci->param_count, fci_cache->called_scope, fci_cache->object);
```
some tests are changed from Error to OK and from OK to Error
My favorite tests to check are ModelsForeignKeysTest and FormsTest

SegmentationFault are on get_called_class is because when get EG(current_execute_data)->This have problem with scope name (zend_string*) to not assigned in memory
This error is not always
example i get this when Phalcon\Di->set

Where should pay attention
->This and EG(current_execute_data)

**Comment by Jurigag** [comment]
No problem. We still need to wait for @andresgutierrez to review this or someone else.

**Comment by andresgutierrez** [comment]
I'm trying a different approach and will review this in a few days.

**Comment by dreamsxin** [comment]
What is the progress of the work?

**Comment by therosco** [comment]
Any progress?

**Comment by sergeyklay** [comment]
Cc: @andresgutierrez

**Comment by sergeyklay** [comment]
@rafmichalak84 @dreamsxin I'm going to merge this PR. So I need your opinion guys. Is it ready to merge or not?

**Comment by dreamsxin** [comment]
@sergeyklay When the Phalcon tests it's passed,  It can be merge.

**Comment by Jurigag** [comment]
It can't be cuz it's not making phalcon work with php 7.1 fully @andresgutierrez

**Comment by sergeyklay** [comment]
I ve created test environment: https://github.com/phalcon/dockerfiles/tree/master/php/7.1-dev with cppcheck, gdb, valgrind, etc.

So I'm able to compile Phalcon. But php -m produces segfault:
```
Reading symbols from /usr/local/bin/php...(no debugging symbols found)...done.
[New LWP 20010]
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
Core was generated by `php -m'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x0000562fbe90c980 in zend_do_implement_interface ()

#0  0x0000562fbe90c980 in zend_do_implement_interface ()
No symbol table info available.
#1  0x0000562fbe8ce108 in zend_class_implements ()
No symbol table info available.
#2  0x00007fad22e79744 in zephir_Phalcon_Mvc_Model_Query_Builder_init (type=type@entry=1, module_number=module_number@entry=33) at /app/ext/phalcon/mvc/model/query/builder.zep.c:99
No locals.
#3  0x00007fad22c4c9d2 in zm_startup_phalcon (type=1, module_number=33) at /app/ext/phalcon.c:712
No locals.
#4  0x0000562fbe8cbc0b in zend_startup_module_ex ()
No symbol table info available.
#5  0x0000562fbe8cbcec in ?? ()
No symbol table info available.
#6  0x0000562fbe8d9bcc in zend_hash_apply ()
No symbol table info available.
#7  0x0000562fbe8cbfda in zend_startup_modules ()
No symbol table info available.
#8  0x0000562fbe8532ab in php_module_startup ()
No symbol table info available.
#9  0x0000562fbe97b82d in ?? ()
No symbol table info available.
#10 0x0000562fbe53075b in ?? ()
No symbol table info available.
#11 0x00007fad26a04b45 in __libc_start_main (main=0x562fbe530470, argc=2, argv=0x7ffe56cf3788, init=<optimized out>, fini=<optimized out>, rtld_fini=<optimized out>, stack_end=0x7ffe56cf3778) at libc-start.c:287
        result = <optimized out>
        unwind_buf = {cancel_jmp_buf = {{jmp_buf = {0, -1657490804047050535, 94763056564970, 140730354841472, 0, 0, 1658386221269551321, 1703809427510298841}, mask_was_saved = 0}}, priv = {pad = {0x0, 0x0, 0x562fbe985690 <__libc_csu_init>, 0x7ffe56cf3788}, data = {prev = 0x0, cleanup = 0x0, canceltype = -1097312624}}}
        not_first_call = <optimized out>
#12 0x0000562fbe530b13 in _start ()
No symbol table info available.

rax            0xa00000001	42949672961
rbx            0x562fbf4dad50	94763072990544
rcx            0x562fbf4d8dd8	94763072982488
rdx            0xffffffff	4294967295
rsi            0x562fbf4dad50	94763072990544
rdi            0x562fbf612fc0	94763074269120
rbp            0x562fbf4d8dd0	0x562fbf4d8dd0
rsp            0x7ffe56cf1db0	0x7ffe56cf1db0
r8             0x250	592
r9             0xfffffffffffff9d0	-1584
r10            0xfffffffffffff9c0	-1600
r11            0x7fad26b4ff90	140381655465872
r12            0x562fbf49a710	94763072726800
r13            0x562fbf612fc0	94763074269120
r14            0x562fbf4dac30	94763072990256
r15            0x562fbf612f10	94763074268944
rip            0x562fbe90c980	0x562fbe90c980 <zend_do_implement_interface+496>
eflags         0x10202	[ IF RF ]
cs             0x33	51
ss             0x2b	43
ds             0x0	0
es             0x0	0
fs             0x0	0
gs             0x0	0

=> 0x562fbe90c980 <zend_do_implement_interface+496>:	addl   $0x1,(%rax)
   0x562fbe90c983 <zend_do_implement_interface+499>:	mov    $0x20,%edi
   0x562fbe90c988 <zend_do_implement_interface+504>:	callq  0x562fbe8967c0 <__zend_malloc>
   0x562fbe90c98d <zend_do_implement_interface+509>:	mov    (%r12),%rdx
   0x562fbe90c991 <zend_do_implement_interface+513>:	mov    %rdx,(%rax)
   0x562fbe90c994 <zend_do_implement_interface+516>:	mov    0x8(%r12),%rdx
   0x562fbe90c999 <zend_do_implement_interface+521>:	mov    %rdx,0x8(%rax)
   0x562fbe90c99d <zend_do_implement_interface+525>:	mov    0x10(%r12),%rdx
   0x562fbe90c9a2 <zend_do_implement_interface+530>:	mov    %rdx,0x10(%rax)
   0x562fbe90c9a6 <zend_do_implement_interface+534>:	mov    0x18(%r12),%rdx
   0x562fbe90c9ab <zend_do_implement_interface+539>:	mov    %rdx,0x18(%rax)
   0x562fbe90c9af <zend_do_implement_interface+543>:	mov    0x18(%rsp),%rdx
   0x562fbe90c9b4 <zend_do_implement_interface+548>:	mov    %rbx,%rsi
   0x562fbe90c9b7 <zend_do_implement_interface+551>:	mov    %r13,%rdi
   0x562fbe90c9ba <zend_do_implement_interface+554>:	mov    %rax,0x20(%rsp)
   0x562fbe90c9bf <zend_do_implement_interface+559>:	movl   $0x11,0x28(%rsp)

Thread 1 (Thread 0x7fad291d9800 (LWP 20010)):
#0  0x0000562fbe90c980 in zend_do_implement_interface ()
#1  0x0000562fbe8ce108 in zend_class_implements ()
#2  0x00007fad22e79744 in zephir_Phalcon_Mvc_Model_Query_Builder_init (type=type@entry=1, module_number=module_number@entry=33) at /app/ext/phalcon/mvc/model/query/builder.zep.c:99
#3  0x00007fad22c4c9d2 in zm_startup_phalcon (type=1, module_number=33) at /app/ext/phalcon.c:712
#4  0x0000562fbe8cbc0b in zend_startup_module_ex ()
#5  0x0000562fbe8cbcec in ?? ()
#6  0x0000562fbe8d9bcc in zend_hash_apply ()
#7  0x0000562fbe8cbfda in zend_startup_modules ()
#8  0x0000562fbe8532ab in php_module_startup ()
#9  0x0000562fbe97b82d in ?? ()
#10 0x0000562fbe53075b in ?? ()
#11 0x00007fad26a04b45 in __libc_start_main (main=0x562fbe530470, argc=2, argv=0x7ffe56cf3788, init=<optimized out>, fini=<optimized out>, rtld_fini=<optimized out>, stack_end=0x7ffe56cf3778) at libc-start.c:287
#12 0x0000562fbe530b13 in _start ()
```

**Comment by sergeyklay** [comment]
@rafmichalak84 @dreamsxin Thank you for contributing!

**Comment by Jurigag** [comment]
@rafmichalak84 @dreamsxin 

There is issue when calling static method from non-static method:

```zep
class Test2
{
	public function testCalled()
	{

		self::testCalled3();
	}
	
	public static function testCalled2()
	{
		self::testCalled3();
	}
	
	protected static function testCalled3()
	{
		var_dump(get_called_class());
	}
}
```

And basically:

```
$test = new Test();
$test->testCalled2(); // works fine
$test->testCalled(); // seg fault
```

Just some scope issue, idk really, before this PR it's broken even more :(

Just let's move to https://github.com/phalcon/zephir/issues/1392

**Pull request merged**